Incompatibiliy between `axum_test::TestServer` and `tower_http::NormalizePathLayer`

I think it is not possible to instantiate a [TestServer](https://docs.rs/axum-test/latest/axum_test/struct.TestServer.html) that also applies path normalization (trailing slash trimming).

Prior discussion: https://github.com/tokio-rs/axum/discussions/2833

I tried this code:

```
let router = Router::new();
let app = NormalizePathLayer::trim_trailing_slash().layer(router);
let app = ServiceExt::<axum::extract::Request>::into_make_service(app);

let server = TestServer::new(app).unwrap();
```

but it fails to compile and I cannot see a way around it:

```
error[E0277]: the trait bound `axum::routing::IntoMakeService<tower_http::normalize_path::NormalizePath<axum::Router>>: axum_test::transport_layer::IntoTransportLayer` is not satisfied
 --> src\server\axum\test.rs:40:32
  |
40 |         let server = TestServer::new(app).unwrap();
  |                      --------------- ^^^ the trait `axum_test::transport_layer::IntoTransportLayer` is not implemented for `axum::routing::IntoMakeService<tower_http::normalize_path::NormalizePath<axum::Router>>`
  |                      |
  |                      required by a bound introduced by this call
  |
  = help: the trait `axum_test::transport_layer::IntoTransportLayer` is implemented for `axum::routing::IntoMakeService<axum::Router>
```

Thanks in advance!

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,7 +1,7 @@
 [package]
 name = "axum-test"
 authors = ["Joseph Lenton <josephlenton@gmail.com>"]
-version = "15.6.0"
+version = "15.7.0"
 rust-version = "1.75"
 edition = "2021"
 license = "MIT"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -45,7 +45,7 @@ serde = { version = "1.0" }
 serde_json = "1.0"
 serde_urlencoded = "0.7"
 smallvec = "1.13"
-tokio = { version = "1.39", features = ["rt", "time"] }
+tokio = { version = "1.40", features = ["rt", "time"] }
 tower = { version = "0.5", features = ["util", "make"] }
 url = "2.5"
 
@@ -79,4 +79,5 @@ regex = "1.10"
 serde-email = { version = "3.0", features = ["serde"] }
 shuttle-axum = "0.47"
 shuttle-runtime = "0.47"
-tokio = { version = "1.39", features = ["rt", "rt-multi-thread", "time", "macros"] }
+tokio = { version = "1.40", features = ["rt", "rt-multi-thread", "time", "macros"] }
+tower-http = { version = "0.5", features = ["normalize-path"] }

--- a/src/internals/transport_layer/mock_transport_layer.rs
+++ b/src/internals/transport_layer/mock_transport_layer.rs
@@ -1,7 +1,7 @@
 use ::anyhow::Error as AnyhowError;
 use ::anyhow::Result;
 use ::axum::body::Body;
-use ::axum::Router;
+use ::axum::response::Response as AxumResponse;
 use ::bytes::Bytes;
 use ::http::Request;
 use ::http::Response;
@@ -18,22 +18,25 @@ pub struct MockTransportLayer<S> {
     service: S,
 }
 
-impl<S> MockTransportLayer<S>
+impl<S, RouterService> MockTransportLayer<S>
 where
-    S: Service<Request<Body>, Response = Router> + Clone + Send + Sync,
+    S: Service<Request<Body>, Response = RouterService> + Clone + Send,
     AnyhowError: From<S::Error>,
     S::Future: Send,
+    RouterService: Service<Request<Body>, Response = AxumResponse>,
 {
     pub(crate) fn new(service: S) -> Self {
         Self { service }
     }
 }
 
-impl<S> TransportLayer for MockTransportLayer<S>
+impl<S, RouterService> TransportLayer for MockTransportLayer<S>
 where
-    S: Service<Request<Body>, Response = Router> + Clone + Send + Sync,
+    S: Service<Request<Body>, Response = RouterService> + Clone + Send,
     AnyhowError: From<S::Error>,
     S::Future: Send,
+    RouterService: Service<Request<Body>, Response = AxumResponse>,
+    AnyhowError: From<RouterService::Error>,
 {
     fn send<'a>(
         &'a self,
@@ -49,7 +52,6 @@ where
             let router = service.oneshot(empty_request).await?;
 
             let response = router.oneshot(request).await?;
-
             Ok(response)
         })
     }

--- a/src/transport_layer/into_transport_layer.rs
+++ b/src/transport_layer/into_transport_layer.rs
@@ -3,6 +3,8 @@ use ::anyhow::Result;
 use crate::transport_layer::TransportLayer;
 use crate::transport_layer::TransportLayerBuilder;
 
+// mod into_make_service_tower;
+
 mod into_make_service;
 mod into_make_service_with_connect_info;
 mod router;

--- a/src/transport_layer/into_transport_layer/into_make_service.rs
+++ b/src/transport_layer/into_transport_layer/into_make_service.rs
@@ -1,6 +1,9 @@
 use ::anyhow::Result;
+use ::axum::extract::Request as AxumRequest;
+use ::axum::response::Response as AxumResponse;
 use ::axum::routing::IntoMakeService;
-use ::axum::Router;
+use ::std::convert::Infallible;
+use ::tower::Service;
 use ::url::Url;
 
 use crate::internals::HttpTransportLayer;
@@ -10,7 +13,11 @@ use crate::transport_layer::TransportLayer;
 use crate::transport_layer::TransportLayerBuilder;
 use crate::util::spawn_serve;
 
-impl IntoTransportLayer for IntoMakeService<Router> {
+impl<S> IntoTransportLayer for IntoMakeService<S>
+where
+    S: Service<AxumRequest, Response = AxumResponse, Error = Infallible> + Clone + Send + 'static,
+    S::Future: Send,
+{
     fn into_http_transport_layer(
         self,
         builder: TransportLayerBuilder,
@@ -37,12 +44,14 @@ impl IntoTransportLayer for IntoMakeService<Router> {
 
 #[cfg(test)]
 mod test_into_http_transport_layer_for_into_make_service {
+    use crate::TestServerConfig;
+    use ::axum::extract::Request;
     use ::axum::extract::State;
     use ::axum::routing::get;
-    use ::axum::routing::IntoMakeService;
     use ::axum::Router;
-
-    use crate::TestServerConfig;
+    use ::axum::ServiceExt;
+    use ::tower::Layer;
+    use ::tower_http::normalize_path::NormalizePathLayer;
 
     async fn get_ping() -> &'static str {
         "pong!"
@@ -55,7 +64,7 @@ mod test_into_http_transport_layer_for_into_make_service {
     #[tokio::test]
     async fn it_should_create_and_test_with_make_into_service() {
         // Build an application with a route.
-        let app: IntoMakeService<Router> = Router::new()
+        let app = Router::new()
             .route("/ping", get(get_ping))
             .into_make_service();
 
@@ -72,10 +81,29 @@ mod test_into_http_transport_layer_for_into_make_service {
     #[tokio::test]
     async fn it_should_create_and_test_with_make_into_service_with_state() {
         // Build an application with a route.
-        let app: IntoMakeService<Router> = Router::new()
+        let app = Router::new()
             .route("/count", get(get_state))
             .with_state(123)
             .into_make_service();
+
+        // Run the server.
+        let server = TestServerConfig::builder()
+            .http_transport()
+            .build_server(app)
+            .expect("Should create test server");
+
+        // Get the request.
+        server.get(&"/count").await.assert_text(&"count is 123");
+    }
+
+    #[tokio::test]
+    async fn it_should_create_and_run_with_router_wrapped_service() {
+        // Build an application with a route.
+        let router = Router::new()
+            .route("/count", get(get_state))
+            .with_state(123);
+        let normalized_router = NormalizePathLayer::trim_trailing_slash().layer(router);
+        let app = ServiceExt::<Request>::into_make_service(normalized_router);
 
         // Run the server.
         let server = TestServerConfig::builder()
@@ -90,12 +118,15 @@ mod test_into_http_transport_layer_for_into_make_service {
 
 #[cfg(test)]
 mod test_into_mock_transport_layer_for_into_make_service {
+    use crate::TestServerConfig;
+    use ::axum::extract::Request;
     use ::axum::extract::State;
     use ::axum::routing::get;
     use ::axum::routing::IntoMakeService;
     use ::axum::Router;
-
-    use crate::TestServerConfig;
+    use ::axum::ServiceExt;
+    use ::tower::Layer;
+    use ::tower_http::normalize_path::NormalizePathLayer;
 
     async fn get_ping() -> &'static str {
         "pong!"
@@ -129,6 +160,25 @@ mod test_into_mock_transport_layer_for_into_make_service {
             .route("/count", get(get_state))
             .with_state(123)
             .into_make_service();
+
+        // Run the server.
+        let server = TestServerConfig::builder()
+            .mock_transport()
+            .build_server(app)
+            .expect("Should create test server");
+
+        // Get the request.
+        server.get(&"/count").await.assert_text(&"count is 123");
+    }
+
+    #[tokio::test]
+    async fn it_should_create_and_run_with_router_wrapped_service() {
+        // Build an application with a route.
+        let router = Router::new()
+            .route("/count", get(get_state))
+            .with_state(123);
+        let normalized_router = NormalizePathLayer::trim_trailing_slash().layer(router);
+        let app = ServiceExt::<Request>::into_make_service(normalized_router);
 
         // Run the server.
         let server = TestServerConfig::builder()

--- a/src/transport_layer/into_transport_layer/into_make_service_with_connect_info.rs
+++ b/src/transport_layer/into_transport_layer/into_make_service_with_connect_info.rs
@@ -1,8 +1,11 @@
 use ::anyhow::anyhow;
 use ::anyhow::Result;
 use ::axum::extract::connect_info::IntoMakeServiceWithConnectInfo;
+use ::axum::extract::Request as AxumRequest;
+use ::axum::response::Response as AxumResponse;
 use ::axum::serve::IncomingStream;
-use ::axum::Router;
+use ::std::convert::Infallible;
+use ::tower::Service;
 use ::url::Url;
 
 use crate::internals::HttpTransportLayer;
@@ -11,9 +14,11 @@ use crate::transport_layer::TransportLayer;
 use crate::transport_layer::TransportLayerBuilder;
 use crate::util::spawn_serve;
 
-impl<C> IntoTransportLayer for IntoMakeServiceWithConnectInfo<Router, C>
+impl<S, C> IntoTransportLayer for IntoMakeServiceWithConnectInfo<S, C>
 where
     for<'a> C: axum::extract::connect_info::Connected<IncomingStream<'a>>,
+    S: Service<AxumRequest, Response = AxumResponse, Error = Infallible> + Clone + Send + 'static,
+    S::Future: Send,
 {
     fn into_http_transport_layer(
         self,
@@ -47,11 +52,14 @@ where
 
 #[cfg(test)]
 mod test_into_http_transport_layer_for_into_make_service_with_connect_info {
+    use crate::TestServerConfig;
+    use ::axum::extract::Request;
     use ::axum::routing::get;
     use ::axum::Router;
+    use ::axum::ServiceExt;
     use ::std::net::SocketAddr;
-
-    use crate::TestServerConfig;
+    use ::tower::Layer;
+    use ::tower_http::normalize_path::NormalizePathLayer;
 
     async fn get_ping() -> &'static str {
         "pong!"
@@ -63,6 +71,25 @@ mod test_into_http_transport_layer_for_into_make_service_with_connect_info {
         let app = Router::new()
             .route("/ping", get(get_ping))
             .into_make_service_with_connect_info::<SocketAddr>();
+
+        // Run the server.
+        let server = TestServerConfig::builder()
+            .http_transport()
+            .build_server(app)
+            .expect("Should create test server");
+
+        // Get the request.
+        server.get(&"/ping").await.assert_text(&"pong!");
+    }
+
+    #[tokio::test]
+    async fn it_should_create_and_run_with_router_wrapped_service() {
+        // Build an application with a route.
+        let router = Router::new().route("/ping", get(get_ping));
+        let normalized_router = NormalizePathLayer::trim_trailing_slash().layer(router);
+        let app = ServiceExt::<Request>::into_make_service_with_connect_info::<SocketAddr>(
+            normalized_router,
+        );
 
         // Run the server.
         let server = TestServerConfig::builder()

--- a/src/transport_layer/transport_layer.rs
+++ b/src/transport_layer/transport_layer.rs
@@ -9,7 +9,7 @@ use ::url::Url;
 
 use crate::transport_layer::TransportLayerType;
 
-pub trait TransportLayer: Debug + Send + Sync {
+pub trait TransportLayer: Debug + Send {
     fn send<'a>(
         &'a self,
         request: Request<Body>,

--- a/src/util/spawn_serve.rs
+++ b/src/util/spawn_serve.rs
@@ -9,7 +9,7 @@ use ::tower::Service;
 
 use crate::util::ServeHandle;
 
-/// A wrapper around [`axum::serve`] for tests,
+/// A wrapper around [`axum::serve()`] for tests,
 /// which spawns the service in a new thread.
 ///
 /// The [`crate::util::ServeHandle`] returned will automatically attempt

